source/gloo.go

```diff
@@ -0,0 +1,200 @@
+		proxy := proxy{}
+		jsonString, err := obj.MarshalJSON()
+		if err != nil {
+			return nil, err
+		}
+		err = json.Unmarshal(jsonString, &proxy)
+		if err != nil {
+			return nil, err
+		}
+		log.Debugf("Gloo: Find %s proxy", proxy.Metadata.Name)
+		proxyTargets, err := gs.proxyTargets(ctx, proxy.Metadata.Name)
+		if err != nil {
+			return nil, err
+		}
+		log.Debugf("Gloo[%s]: Find %d target(s) (%+v)", proxy.Metadata.Name, len(proxyTargets), proxyTargets)
+		proxyEndpoints, err := gs.generateEndpointsFromProxy(ctx, &proxy, proxyTargets)
+		if err != nil {
+			return nil, err
+		}
+		log.Debugf("Gloo[%s]: Generate %d endpoint(s)", proxy.Metadata.Name, len(proxyEndpoints))
+		endpoints = append(endpoints, proxyEndpoints...)
+	}
+	return endpoints, nil
+}
+
+func (gs *glooSource) generateEndpointsFromProxy(ctx context.Context, proxy *proxy, targets endpoint.Targets) ([]*endpoint.Endpoint, error) {
+	endpoints := []*endpoint.Endpoint{}
+	for _, listener := range proxy.Spec.Listeners {
+		for _, virtualHost := range listener.HTTPListener.VirtualHosts {
+			annotations, err := gs.annotationsFromProxySource(ctx, virtualHost)
+			if err != nil {
+				return nil, err
+			}
+			ttl, err := getTTLFromAnnotations(annotations)
+			if err != nil {
+				return nil, err
+			}
+			providerSpecific, setIdentifier := getProviderSpecificAnnotations(annotations)
+			for _, domain := range virtualHost.Domains {
+				endpoints = append(endpoints, endpointsForHostname(strings.TrimSuffix(domain, "."), targets, ttl, providerSpecific, setIdentifier)...)
+			}
+		}
+	}
+	return endpoints, nil
+}
+
+func (gs *glooSource) annotationsFromProxySource(ctx context.Context, virtualHost proxyVirtualHost) (map[string]string, error) {
+	annotations := map[string]string{}
+	for _, src := range virtualHost.Metadata.Source {
+		kind := sourceKind(src.Kind)
+		if kind != nil {
+			source, err := gs.dynamicKubeClient.Resource(*kind).Namespace(src.Namespace).Get(ctx, src.Name, metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+			for key, value := range source.GetAnnotations() {
+				annotations[key] = value
+			}
+		}
+	}
+	return annotations, nil
+}
+
+func (gs *glooSource) proxyTargets(ctx context.Context, name string) (endpoint.Targets, error) {
+	svc, err := gs.kubeClient.CoreV1().Services(gs.glooNamespace).Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	var targets endpoint.Targets
+	switch svc.Spec.Type {
+	case corev1.ServiceTypeLoadBalancer:
+		for _, lb := range svc.Status.LoadBalancer.Ingress {
+			if lb.IP != "" {
+				targets = append(targets, lb.IP)
+			}
+			if lb.Hostname != "" {
+				targets = append(targets, lb.Hostname)
+			}
+		}
+	default:
+		log.WithField("gateway", name).WithField("service", svc).Warn("Gloo: Proxy service type not supported")
+	}
+	return targets, nil
+}
+
+func sourceKind(kind string) *schema.GroupVersionResource {
+	switch kind {
+	case "*v1.VirtualService":
```

Comment:
Is this right?

Reply:
Yep, even if it's seem funky.
In a route definition you have "source" to retrace back user create resource. (To get any annotations for options/provider specific) 
```
        routes:
        - matchers:
          - prefix: /
          metadata:
            sources:
            - kind: '*v1.VirtualService'
              name: grafana-https
              namespace: monitor
              observedGeneration: 147
```
For that I need to map the "kind" to a k8s resource type, to get metadata.
And the switch is here in case we want to add new resource types

I didn't find a better way to do it, if you have any idea to improve that, I'll take it :)